src/acl/AtStep.cc

```diff
@@ -8,25 +8,40 @@
 
 #include "squid.h"
 
-#if USE_OPENSSL
-
 #include "acl/AtStep.h"
 #include "acl/AtStepData.h"
 #include "acl/FilledChecklist.h"
 #include "client_side.h"
 #include "http/Stream.h"
+#if USE_OPENSSL
 #include "ssl/ServerBump.h"
+#endif
 
 int
-ACLAtStepStrategy::match (ACLData<Ssl::BumpStep> * &data, ACLFilledChecklist *checklist)
+ACLAtStepStrategy::match(ACLData<XactionStep> * &data, ACLFilledChecklist *checklist)
 {
+#if USE_OPENSSL
     Ssl::ServerBump *bump = NULL;
-    if (checklist->conn() != NULL && (bump = checklist->conn()->serverBump()))
-        return data->match(bump->step);
-    else
-        return data->match(Ssl::bumpStep1);
-    return 0;
-}
+    if (checklist->conn() != NULL &&
+        (bump = checklist->conn()->serverBump()) &&
+        data->match(bump->step))
+        return 1;
+#endif
+
+    if (checklist->request && data->match(XactionStep::generatingConnect)) {
+        if (!checklist->request->masterXaction)
+            debugs(28, DBG_IMPORTANT, "at_step GeneratingCONNECT ACL is missing master transaction info. Assuming mismatch.");
```

Comment:
The `Assuming mismatch` part a lie -- the code may still return 1 which means it does not simply "assume mismatch". If this check cannot be moved lower, then we have to remember that `generatingConnectMatchedWithoutMx`, wait until we know the outcome of the tlsBump1 match() below, and only then, _if_ that outcome is falsy and generatingConnectMatchedWithoutMx is true, display this warning (and return 0).

The above suggestion assumes that the match(x) and match(y) can both report a match for the same `data` object. Otherwise, it does not make sense to wait until that second match() call before returning 0.

BTW, please prefix this with `WARNING: `, for consistency (and clarity) sake.

Reply:
> The Assuming mismatch part a lie -- the code may still return 1 which means it does not simply "assume mismatch".

I saw it but I consider it no huge problem, because the generatingConnect may missmatch but the step1 may match, in the case they are both used at the same acl line, but the admin still needs to now that the generatingConnect is problematic here.

Anyway the code is fixed now.

> If this check cannot be moved lower, 

The check can be moved under the check of tlsBump1 I was wrong about. Now the related code does the following:
 - If there is a ServerBump structure retrieve the bumping step from here.  If the acl matches the current bumping step return 1;
 - else assume that we are at bumping step1 and check if acl matches the bumping step1
 - else check if the acl matches the generatingConnect. etc.

Commit ece98b98.